plutus-core/plutus-core/src/PlutusCore/Name/UniqueSet.hs

```diff
@@ -0,0 +1,91 @@
+module PlutusCore.Name.UniqueSet (
+  UniqueSet (..),
+  insertByUnique,
+  insertByName,
+  singletonName,
+  fromFoldable,
+  fromUniques,
+  fromNames,
+  memberByUnique,
+  memberByName,
+  notMemberByName,
+  (\\),
+  union,
+  setOfByUnique,
+  setOfByName,
+) where
+
+import Control.Lens (Getting, view)
+import Control.Lens.Getter (views)
+import Data.Coerce (Coercible, coerce)
+import Data.Foldable (foldl')
+import Data.IntSet qualified as IS
+import Data.IntSet.Lens qualified as IS
+import PlutusCore.Name.Unique (HasUnique (..), Unique (Unique))
+
+-- | A set containing uniques.
+newtype UniqueSet unique = UniqueSet
+  { unUniqueSet :: IS.IntSet
+  }
+  deriving newtype (Show, Eq, Semigroup, Monoid)
+
+-- | Insert a value by a unique.
+insertByUnique ::
+  (Coercible unique Unique) =>
+  unique ->
+  UniqueSet unique ->
+  UniqueSet unique
+insertByUnique = coerce . IS.insert . coerce
+
+-- | Insert a value by the unique of a name.
+insertByName :: (HasUnique name unique) => name -> UniqueSet unique -> UniqueSet unique
+insertByName = insertByUnique . view unique
+
+singletonName :: (HasUnique name unique) => name -> UniqueSet unique
+singletonName n = insertByName n mempty
+
+-- | Convert a 'Foldable' into a 'UniqueSet' using the given insertion function.
+fromFoldable ::
```

Comment:
also kind of dubious about this. I would just do something like `foldMap (singleton . coerce)` or something

Reply:
Why is this dubious?